Library/Source/ShaderCompilerOpenGL.cpp

```diff
@@ -1,19 +1,164 @@
+
+            auto compiler = std::make_unique<spirv_cross::CompilerGLSL>(parser.get_parsed_ir());
+            
+            compiler->build_combined_image_samplers();
+            
+            spirv_cross::CompilerGLSL::Options options = compiler->get_common_options();
+
+#ifdef ANDROID
+            options.version = 300;
+            options.es = true;
+#else
+            options.version = 430;
+            options.es = false;
+#endif
+            compiler->set_common_options(options);
+
+            // rebuild uniform lists with the correct type
+            program.buildReflection();
+            int numUniforms = program.getNumUniformVariables();
+
+            for (int i = 0; i < numUniforms; i++)
+            {
+                const auto& uniform = program.getUniform(i);
+                std::string uniformString = "uniform highp ";
+                if (uniform.glDefineType == 0x8B5C) //GL_FLOAT_MAT4
+                {
+                    uniformString += "mat4 ";
+                }
+                else
+                {
+                    uniformString += "vec4 ";
+                }
+                uniformString += uniform.name;
+                uniformString += ";";
+                compiler->add_header_line(uniformString);
+            }
+
+            std::string compiled = compiler->compile();
+
+            // rename "uniform Frame { .... }" . Keep the uniforms
+            const std::string frameNewName = ((stage == EShLangVertex) ? "FrameVS" : "FrameFS");
+            const std::string frame = "Frame";
+            size_t pos = compiled.find(frame);
+            if (pos != std::string::npos)
+            {
+                compiled.replace(pos, frame.size(), frameNewName);
+            }
+            
+            spirv_cross::ShaderResources resources = compiler->get_shader_resources();
+            for (auto& resource : resources.uniform_buffers)
+            {
+                auto str = resource.name;
+                auto fbn = compiler->get_fallback_name(resource.id);
+                std::string rep = fbn + ".";
+                size_t pos = compiled.find(rep);
+                while (pos != std::string::npos)
+                {
+                    compiled.replace(pos, rep.size(), "");
+                    pos = compiled.find(rep);
+                }
+            }
+
+#ifdef ANDROID
+            glsl = compiled.substr(strlen("#version 300 es\n"));
+
+            // frag def
+            static const std::string fragDef = "layout(location = 0) out highp vec4 glFragColor;";
+            pos = glsl.find(fragDef);
+            if (pos != std::string::npos)
+            {
+                glsl.replace(pos, fragDef.size(), "");
+            }
+
+            // frag
+            static const std::string fragColor = "glFragColor";
+            pos = glsl.find(fragColor);
+            if (pos != std::string::npos)
+            {
+                glsl.replace(pos, fragColor.size(), "gl_FragColor");
+            }
```

Comment:
What is happening with this and why is it necessary?

Reply:
That's because bgfx changes a couple things on FS outputs : https://github.com/bkaradzic/bgfx/blob/master/src/renderer_gl.cpp#L5842